scripts/gen_layer.py

```diff
@@ -0,0 +1,3723 @@
+                                                                                  VkLayerProperties *pProperties) {
+    return EnumerateInstanceLayerProperties(pCount, pProperties);
+}
+
+VK_LAYER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL vkEnumerateInstanceExtensionProperties(const char *pLayerName, uint32_t *pCount,
+                                                                                      VkExtensionProperties *pProperties) {
+    return EnumerateInstanceExtensionProperties(pLayerName, pCount, pProperties);
+}
+
+VK_LAYER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL vkEnumeratePhysicalDevices(VkInstance instance, uint32_t *pPhysicalDeviceCount,
+                                                                          VkPhysicalDevice *pPhysicalDevices) {
+    return EnumeratePhysicalDevices(instance, pPhysicalDeviceCount, pPhysicalDevices);
+}
+
+VK_LAYER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL vkNegotiateLoaderLayerInterfaceVersion(VkNegotiateLayerInterface *pVersionStruct) {
+    assert(pVersionStruct != NULL);
+    assert(pVersionStruct->sType == LAYER_NEGOTIATE_INTERFACE_STRUCT);
+
+    if (pVersionStruct->loaderLayerInterfaceVersion > CURRENT_LOADER_LAYER_INTERFACE_VERSION) {
+        // Loader is requesting newer interface version; reduce to the version we support.
+        pVersionStruct->loaderLayerInterfaceVersion = CURRENT_LOADER_LAYER_INTERFACE_VERSION;
+    } else if (pVersionStruct->loaderLayerInterfaceVersion < CURRENT_LOADER_LAYER_INTERFACE_VERSION) {
+        // Loader is requesting older interface version; record the Loader's version
+        loader_layer_iface_version = pVersionStruct->loaderLayerInterfaceVersion;
+    }
+
+    if (pVersionStruct->loaderLayerInterfaceVersion >= 2) {
+        pVersionStruct->pfnGetInstanceProcAddr = vkGetInstanceProcAddr;
+        pVersionStruct->pfnGetDeviceProcAddr = nullptr;
+        pVersionStruct->pfnGetPhysicalDeviceProcAddr = nullptr;
+    }
+
+    return VK_SUCCESS;
+}
+
+// vim: set sw=4 ts=8 et ic ai:
+'''
+
+class VulkanProfilesLayerGenerator():
+    skipped_extensions = ['VK_KHR_external_memory_capabilities']
+    skipped_exts = ['NVX']
+    emulated_extensions = ['VK_KHR_portability_subset']
+    non_modifiable_structs = ['VkPhysicalDevicePointClippingProperties', 'VkPhysicalDevicePointClippingPropertiesKHR', 'VkPhysicalDeviceDriverProperties', 'VkPhysicalDeviceDriverPropertiesKHR', 'VkPhysicalDeviceIDProperties', 'VkPhysicalDeviceIDPropertiesKHR', 'VkPhysicalDeviceMemoryBudgetPropertiesEXT', 'VkPhysicalDevicePCIBusInfoPropertiesEXT', 'VkPhysicalDeviceDrmPropertiesEXT', 'VkPhysicalDeviceToolProperties', 'VkPhysicalDeviceToolPropertiesEXT', 'VkPhysicalDeviceGroupProperties']
+    additional_features = ['VkPhysicalDeviceFeatures']
+    additional_properties = ['VkPhysicalDeviceProperties', 'VkPhysicalDeviceLimits', 'VkPhysicalDeviceSparseProperties', 'VkPhysicalDeviceToolProperties']
```

Comment:
I guess the reason why there's a need for `additional_features/properties` is because of nested structures.

Is that correct?

Sure, this works, but it's less future proof, so instead maybe parsing recursively for nested structures is better.

Reply:
No, these are structures that do not extend VkPhysicalDeviceProperties2 or VkPhysicalDeviceFeatuers2, but we still have them in the schema.